services/device-registry-mongodb/src/main/java/org/eclipse/hono/deviceregistry/mongodb/service/MongoDbBasedRegistrationService.java

```diff
@@ -179,28 +171,11 @@ public void stop(final Promise<Void> stopPromise) {
         Objects.requireNonNull(deviceId);
         Objects.requireNonNull(resourceVersion);
 
-        if (!config.isModificationEnabled()) {
-            final String errorMsg = String.format("Modification is disabled for tenant [%s]", tenantId);
-            TracingHelper.logError(span, errorMsg);
-            log.debug(errorMsg);
-            return Future.succeededFuture(Result.from(HttpURLConnection.HTTP_FORBIDDEN, OperationResult::empty));
-        }
-
-        return findDevice(tenantId, deviceId)
-                .compose(deviceDto -> {
-                    if (deviceDto == null) {
-                        final String errorMsg = String.format("Device [%s] not found.", deviceId);
-                        log.debug(errorMsg);
-                        TracingHelper.logError(span, errorMsg);
-                        return Future.succeededFuture(OperationResult.empty(HttpURLConnection.HTTP_NOT_FOUND));
-                    }
-                    if (!deviceDto.isVersionMatch(resourceVersion)) {
-                        log.debug("Resource Version mismatch");
-                        TracingHelper.logError(span, "Resource Version mismatch");
-                        return Future.succeededFuture(OperationResult.empty(HttpURLConnection.HTTP_PRECON_FAILED));
-                    }
-                    return processUpdateDevice(tenantId, deviceId, deviceDto, span);
-                });
+        return MongoDbServiceUtils.isAllowedToModify(config, tenantId)
+                .compose(ok -> findDevice(tenantId, deviceId))
+                .compose(deviceDto -> MongoDbServiceUtils.checkResourceVersion(deviceDto, resourceVersion))
+                .compose(deviceDto -> processUpdateDevice(tenantId, deviceDto, span))
+                .recover(error -> Future.succeededFuture(MongoDbServiceUtils.mapErrorToResult(error, span)));
```

Comment:
my understanding of this code is that we first load the entity, then check if the resource version is correct, then modify the entity and then write it back to the database. I am no expert in Mongo but I would assume that you can simply invoke an update operation providing filter criteria in order to do that in one go, can't you?
This approach here also seems to have the problem that after you retrieved the device and verified that the version matches, another update could take place which changes the version *before* you are writing back the object, or does the Mongo DB driver prevent this somehow?

Reply:
I see the point regarding another update between the retrieval of the device and updating the retrieved device. I don't see any explicit locking mechanisms with the library that we use to lock more than one transactions. In that case, we can also include  _version_ in the filter criteria for the update. The update will be performed only when the _version_ also matches or no _version_ is available in the request. The current code for retrieval of the device and the version check before performing update operation could be removed.

In this approach, when no resource(_device_) is updated, there are atleast two possible reasons. Once is "device not found" and the other is "resource version mismatch". In order to report the response code accordingly, a find query need to be executed to retrieve the device followed by version check.


> I see. So let's go with a String valued version field then and do the serialization as part of the setter, ok?

Previously we discussed to make the _version_ in the `BaseDto` class as `String`. This approach somehow forces us to declare the _version_ as `long`. When a request to update a device arrives, it may not contain a version and in that case, we simply update with the given device information and increment the version value by one. In this approach, during the update operation, we don't know what is the version of the resource and hence we cannot manually increment to one unless queried from the database. For that purpose, we could use "$inc" operator from MongoDB. Using this, we could instruct the MongoDB to increment the _version_ field by one during update operation. One prerequisite is, that this field should be a number.

Reply:
I guess that is the reason why the example registry uses UUID for the version in the first place :-) A new UUID can be computed independently from the existing version ...
So where does this leave us? We stick with the UUID or do you want to use the *auto increment* feature of MongoDB?

Reply:
As  I understood from the gitter chat, we would like to stick to passing the `version` using `String` data type. IMHO allowing the _version_ as _string_ and then enforcing that the string should be a valid number doesn't look so nice and also we could avoid the typecasting concerns. Hence I would prefer using UUID.

Reply:
Now I use UUID and pushed the corresponding changes. Pls take a look :-)